chandra_aca/dark_subtract.py

```diff
@@ -0,0 +1,431 @@
+    if dc_img is not None:
+        assert dc_img.shape == (1024, 1024)
+
+    if dc_img is None:
+        dc_img, dc_tccd = get_dark_data(start)
+
+    if t_ccd is None:
+        t_ccd, t_ccd_times = get_tccd_data(start, stop, source=source)
+
+    if aca_images is not None:
+        # If supplied a dictionary of aca images, use that.  Should be keyed by slot.
+        # This is for testing.
+        assert isinstance(aca_images, dict)
+        # For each slot, require aca_image have columns IMG, IMGROW0_8X8, IMGCOL0_8X8, TIME
+        for slot in aca_images:
+            assert "IMG" in aca_images[slot].colnames
+            assert "IMGROW0_8X8" in aca_images[slot].colnames
+            assert "IMGCOL0_8X8" in aca_images[slot].colnames
+            assert "TIME" in aca_images[slot].colnames
+            # require that IMGROW0_8X8 and IMGCOL0_8X8 are masked columns
+            assert hasattr(aca_images[slot]["IMGROW0_8X8"], "mask")
+            assert hasattr(aca_images[slot]["IMGCOL0_8X8"], "mask")
+        imgs = aca_images.copy()
+    elif source == "maude":
+        imgs = get_maude_images(start, stop)
+    elif source == "mica":
+        imgs = get_mica_images(start, stop)
+    else:
+        raise ValueError(
+            "aca_image_table must be defined or source must be maude or mica"
+        )
+
+    imgs_bgsub = {}
+    dark_imgs = {}
+    for slot in range(8):
+        if slot not in imgs:
+            continue
+        dark_imgs[slot] = get_dark_current_imgs(
+            imgs[slot], dc_img, dc_tccd, t_ccd, t_ccd_times
+        )
+        img_col = "IMG"
+        imgs_bgsub[slot] = imgs[slot][img_col] - dark_imgs[slot]
+        imgs_bgsub[slot].clip(0, None)
+        if full_table:
+            imgs[slot]["IMGBGSUB"] = imgs_bgsub[slot]
+            imgs[slot]["IMGDARK"] = dark_imgs[slot]
+            imgs[slot]["AACCCDPT"] = np.interp(imgs[slot]["TIME"], t_ccd_times, t_ccd)
+
+    if not full_table:
+        return imgs_bgsub
+    else:
+        return imgs
+
+
+def get_dark_current_imgs(img_table, dc_img, dc_tccd, t_ccd, t_ccd_times):
+    """
+    Get the scaled dark current values for a table of ACA images.
+
+    Parameters
+    ----------
+    img_table : astry.table.Table
+        A table containing information about the images.
+    dc_img : 1024x1024 array
+        The dark calibration image.
+    dc_tccd : float
+        The reference temperature of the dark calibration image.
+    t_ccd : array
+        The temperature values covering the time range of the img_table.
+    t_ccd_times : type
+        The corresponding times for the temperature values.
+
+    Returns
+    -------
+    dark : np.array
+        An array containing the temperature scaled dark current for each ACA image
+        in the img_table in e-/s.
+
+    """
+    assert dc_img.shape == (1024, 1024)
+    assert len(t_ccd) == len(t_ccd_times)
+
+    dark_raw = get_dark_backgrounds(
+        dc_img,
+        img_table["IMGROW0_8X8"].data.filled(1025),
```

Comment:
This needs a test that it is doing something reasonable for out-of-bounds data.

Reply:
Not only that, AACCCDPT comes every 32.8 seconds in the best of cases, and I don't see code to guarantee it doesn't. You can try:
```
dark_subtract.dark_subtract.get_tccd_data('2022:001:00:00:10', '2022:001:00:00:42', source="maude")
```

Reply:
I think there's a word missing after "guarantee it doesn't" in the comment but yes it is a good point that for small time ranges this might have unexpected fails on AACCCDPT.  I was think that the standard should probably be to fetch at least a chunk of n minutes and use some default smoothing.

Reply:
Not necessarily a good idea to assume how the code will be used. If there are restrictions, the code should explicitly enforce them or do some kind of check. When running aca_view in real-time mode, each chunk is a few seconds long.

At the very least there should be a check that t_ccd data is not empty and raise an error explaining this.

On possibility is to add some padding in the time when calling this function to guarantee there is data.

Reply:
"One possibility is to add some padding in the start/stop times when calling this function to guarantee there is data." That's what I meant by "fetch at least a chunk of n minutes".

Reply:
yes, but I meant it _inside_ `_get_dcsub_aca_images` so a user does not have to do it, and so you don't get all image telemetry for tima you do not want it.

The user calls `_get_dcsub_aca_images`, in there, you fetch AACCCDPT for a larger interval. All the rest is not padded.

Reply:
"yes, but I meant it inside _get_dcsub_aca_images so a user does not have to do it, and so you don't get all image telemetry for times you do not want."  Yes, that's what I meant too.

Reply:
Since we were talking about fetching AACCCDPT I was talking about fetching at least a certain amount of time of that telemetry and possibly smoothing it as the standard.